e3db/client.py

```diff
@@ -1409,3 +1410,119 @@ def read_file(self, record_id, destination_filename):
+        public_key = note.get_writer_encryption_key()
+        public_signing_key = note.get_writer_signing_key()
+        # decrypt the eak to get ak
+        ak = Crypto.decrypt_note_eak(private_key, eak, public_key)
+        decrypted_note = self.decrypt_note_with_key(note, ak, public_signing_key)
+        # if the note isn't signed, throw an error
+        if decrypted_note.get_signature() == '':
+            raise NoteValidationError('Decrypted note does not include a signature.')
+        return decrypted_note
+
+    @classmethod
+    def decrypt_note_with_key(self, encrypted_note, ak, verifying_key):
+        """
+        Decrypt and validate every field within a note.
+
+        Parameters
+        ----------
+        note : e3db.Note
+            Encrypted note with signed data
+        ak :
+            Raw access key which is used for decryption
+        verifying_key : 
+            Base64 encoded public signing key which is used for validation
+
+        Returns
+        -------
+        e3db.Note
+            Decrypted note
+        """
+
+        # verify the signature from the note
+        verified_salt = self.verify_field('signature', encrypted_note.get_signature(), verifying_key)
+        if verified_salt == encrypted_note.get_signature():
+            signature_salt = None
+        else:
+            signature_salt = verified_salt
+        # copy all of the contents from the encrypted note into a new decrypted note
+        decrypted_note = copy.deepcopy(encrypted_note)
+        decrypted_data = {}
+        # decrypt and verify the signature of each field in data
+        for key in encrypted_note.data:
+            raw_field = Crypto.decrypt_field(encrypted_note.data[key], ak)
+            decrypted_data[key] = self.verify_field(key, raw_field, verifying_key, signature_salt)
+        # replace the data in the new note with the decrypted data
+        decrypted_note.data = decrypted_data
+        return decrypted_note
+
+    @classmethod
+    def verify_field(self, key, value, verifying_key, salt = None):
+        """
+        Verify the key, value & the salt, if it's provided using the verifying key
+
+        Parameters
+        ----------
+        key : 
+            The key for the field which corresponds to the value
+        value :
+            Signed string that needs to be validated
+        verifying_key : 
+            Base64 encoded public signing key which is used for validation
+        salt :
+            A UUID which has already been verified as the salt. If not provided, the salt
+            found in value is used instead
+
+        Returns
+        -------
+        str
+            If verified, the plaintext from the field is returned. Otherwise, an error is thrown.
+        """
+
+        parts = value.split(";")
+        # if the field doesn't have the correct signature version as a prefix, assume it's not signed & return without validating
+        if parts[0] != Crypto.get_signature_version():
```

Comment:
I'd like this (whether to implicitly accept / not raise an error for unsigned notes to be a user controllable option (default to false is fine) as part of the decrypt note method (and ultimately as part of the read note method that will call decrypt note) and trickle down as a parameter to this method

Reply:
added